bloom-housing/bloom
# feat: listings approval emails

# Pull Request Template

## Issue Overview

This PR addresses https://github.com/metrotranscom/doorway/issues/299, https://github.com/metrotranscom/doorway/issues/300,
https://github.com/metrotranscom/doorway/issues/302



## Description

This PR creates a combined endpoint that will update the listing and send a corresponding email to the appropriate users based on the update. The three cases are as follows:
1. A partner submits a draft listing for admin approval. This will send an email to all admin users to review the listing.
2. An admin request changes on a listing. This will send an email to all partners/juris admin that have access to that listing.
3. An admin approves & publishes a listing. This sends an email to all partners/juris admin informing them of a new published listing.

Figma: https://www.figma.com/file/cNRVrLdz8cjXQdSr54BRao/Approval?type=design&node-id=1607-68638&mode=design&t=iL57CewScqDs0pGR-0

Note for Eng: Definitely looking for feedback on my email error handling approach and using the status code of 500 here. I'm feeling unsure if big try/catches is the most elegant solution for this case.

Note for Design (@eajensenwa): I made some guesses on error handling for when the listing is successfully updated but the email fails to send along with when the "broader something went wrong error" occurs but from the details page rather than the edit page.

Currently, the errors are either "Oops! Looks like something went wrong" if the update failed (existing functionality) or "Listing updated but the associated emails failed to send." if just the notify step fails.
<img width="1440" alt="Screenshot 2023-08-28 at 11 58 23 PM" src="https://github.com/bloom-housing/bloom/assets/53269332/b71f5f18-0871-46e2-8a3d-70ca4cf7c06a">

## How Can This Be Tested/Reviewed?

Pull down locally
1. Reseed
4. Sign in to admin@example.com
5. Change Test: Coliseum to a draft listing
6. Create an admin account for your work email and a partner account with another email
7. Sign out
8. Go to work email and confirm + create new account
9. Open new incognito window and confirm + create partner account
11. Go to Test: Coliseum and click submit from partner account
12. Verify that email of your admin account (work email) has received the correct email with functional links mirroring the request approval designs above
14. Request changes from admin account
15. Go to email associated with partner account and verify email accuracy
16. Click submit again from the partners account
17. Switch to the admin account and Approve & Publish
18. Verify the email received by the partner account email

## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
